Fix unit test execution error issue

runchen0919 opened the pull request:
Fixes: https://github.com/eclipseguru/bazel-eclipse/issues/46

cc: @guw

runchen0919 committed:
Fix unit test execution error issue

guw commented on bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainerRuntimeResolver.java:
```diff
@@ -232,6 +232,18 @@ private boolean populateWithSavedContainer(IJavaProject project, ContainerResolu
                     // projects need to be resolved properly so we have all the output folders and exported jars on the classpath
                     var sourceProject = workspaceRoot.getProject(e.getPath().segment(0));
+
+                    // Check for cycles BEFORE calling beginResolvingProject to avoid depth counter issues
```
Hmm, this looks like a bug in `beginResolvingProject`. The counter should probably not be increased

runchen0919 replied:
updated. only increment depth counter for new projects in cycle detection

runchen0919 committed:
fix: only increment depth counter for new projects in cycle detection

Fixed bug where currentDepth was unconditionally incremented in
beginResolvingProject(), causing potential counter mismatches.
Now only increments when project is actually added to processedProjects.

Removed redundant cycle checks in calling code since the method now
handles duplicate projects correctly on its own.

guw commented on bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainerRuntimeResolver.java:
```diff
@@ -68,8 +68,11 @@ public void add(IRuntimeClasspathEntry runtimeClasspathEntry) {
+            if (isNewProject) {
+                currentDepth++;
+            }
+            return isNewProject;
```
Can you rebase once #52 is merged?

runchen0919 replied:
sure, https://github.com/eclipseguru/bazel-eclipse/pull/52 has been merged and I will rebase the main branch.

guw commented on bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainerRuntimeResolver.java:
```diff
@@ -264,9 +267,74 @@ private boolean populateWithSavedContainer(IJavaProject project, ContainerResolu
+        // This ensures that Eclipse-compiled classes (including test classes) are available at runtime
+        addProjectOutputFolders(project, resolutionContext);
+
+        // Note: Test framework dependencies (like JUnit) are now pre-cached in the container during
```
This is weird. Why wouldn't they come from regular Bazel dependencies?

https://github.com/eclipseguru/bazel-eclipse/blob/main/docs/common/classpath.md#basics-classpath

runchen0919 replied:
1. Eclipse compiles to its own output folders:

- Main classes → eclipse-bin
- Test classes → eclipse-testbin

2. Bazel compiles to different locations:

- Bazel outputs go to bazel-bin, bazel-out, etc.

3. The runtime classpath is being built from Bazel's dependency graph (via the saved container), but this graph only knows about:

- JAR dependencies
- Other project dependencies
- Bazel-compiled outputs

4. The project's OWN Eclipse-compiled classes are NOT in Bazel's dependency graph because:

- You don't declare a dependency on yourself in Bazel
- Eclipse's incremental compilation outputs are IDE-specific, not part of Bazel's build model

guw commented on bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainerRuntimeResolver.java:
```diff
@@ -264,9 +267,74 @@ private boolean populateWithSavedContainer(IJavaProject project, ContainerResolu
 
+        // Add the project's own output folders to the runtime classpath
+        // This ensures that Eclipse-compiled classes (including test classes) are available at runtime
+        addProjectOutputFolders(project, resolutionContext);
```
`org.eclipse.jdt.launching.JavaRuntime.resolveRuntimeClasspathEntry(IRuntimeClasspathEntry, IJavaProject, boolean)` would add the output folders alread. I don't think we should be doing this.

I wonder if the error reported in #46 is caused by missing test dependencies.

runchen0919 replied:
When I comment out this line of code, I encounter the following problem.

```
java.lang.NoClassDefFoundError: com/projectName/demo_app/DemoServiceHandler
 at com.projectName.demo_app.DemoServiceHandlerTest.testBarCopiesFieldsToResponses([DemoServiceHandlerTest.java:22](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html))
Caused by: java.lang.ClassNotFoundException: com.projectName.demo_app.DemoServiceHandler
 at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass([BuiltinClassLoader.java:581](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html))
 at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass([ClassLoaders.java:178](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html))
 at java.base/java.lang.ClassLoader.loadClass([ClassLoader.java:527](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html))
 ... 27 more
```
